### PR TITLE
#161342840 Get and Update User Profile

### PR DESCRIPTION
**Pull request by TheoOkafor**
#### What does this PR do?
- Create Get-and-update-user-profile functionality

#### Description of Task to be completed?
- Create the get user profile by username route
- Create the update user profile by username route
- Create utility for checking string length
- Create middleware to handle image upload and storage on cloudinary.com 

#### How should this be manually tested?
- git clone `https://github.com/andela/haven-ah-backend.git`.
- Install `node` if not installed.
- ensure that PostgreSQL is running on your local machine.
- run `npm install` to install the npm packages.
- run `npm test` to run the unit and integration tests.
To test on postman
- run `npm run start-dev`
- On postman run GET `localhost:5000/api/users/:username` (set JWT token as`x-access-token` in header).

#### Any background context you want to provide?
N/A

#### What are the relevant pivotal tracker stories?
#161342851

#### Screenshots (if appropriate)
Good Response:
screen shot 2018-11-01 at 2 07 31 pm

Bad Request Response (when email or username is included):
screen shot 2018-10-31 at 10 48 12 am

When bio exceeds 200 character length:
screen shot 2018-10-31 at 10 49 53 am

### CODE REVIEW

File: server/middlewares/authorized.js [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+ * to complete the request.
+ * @param {object} request
+ * @param {object} response
+ * @param {function} next
+ * @returns {function} callback function
+ */
+const authorization = async (request, response, next) => {
+  const data = await userRepo.getUserbyUsername(request.params.username);
+  const id = data ? data.id : null;
+  request.authorized = id === request.userId; // BOOLEAN
```
Comment (PriscillaSam) [inline server/middlewares/authorized.js]: Can this be renamed like so `request.isAuthorized`?

File: server/tests/controllerTests/user.test.js [hunk 2]
```diff
@@ -124,3 +123,99 @@ describe('GET api/v1/users', () => {
...
+    const response = await chai.request(app)
+      .put('/api/v1/users/i_amsull')
+      .send(goodUserUpdate);
+
+    expect(response.status).to.be.equal(403);
+    expect(response.body.message).to.be.deep
+      .equals('You need to sign in first');
+  });
+
+  it('should return an error if token could not be verified', async () => {
```
Comment (PriscillaSam) [inline server/tests/controllerTests/user.test.js]: Is this the same with the one above? I see an it block with the same description.

File: server/middlewares/authorized.js (hunk 1 above)
Comment (TheoOkafor) [reply server/middlewares/authorized.js]: I will change that now. Thanks.

**Comment by wiztemple** [comment]
201 for update?

**Comment by TheoOkafor** [comment]
> 201 for update?

Noted

**Comment by TheoOkafor** [comment]
I extended the timeout to 25 seconds. This is mostly because of the cloudinary service which takes time.

**Comment by victor-nweze-deel** [comment]
This looks fine. Pls let us sync to integrate your cloudinary function with the saving of articles.

File: package.json [hunk 3]
```diff
@@ -8,13 +8,14 @@
     "babel-server": "babel server/ -d dist/server/ && npm run babel-app",
     "build": "npm run clean && mkdir dist/ && npm run babel-server && npm run babel-app",
     "clean": "rm -rf dist/",
-    "coveralls": "nyc --reporter=lcov --reporter=text --require babel-register  mocha server/tests/**/*.test.js --exit",
-    "cover": "nyc report --reporter=text-lcov | coveralls",
```
Comment (seyi-adeleke) [inline package.json]: Why did you remove the cover script?

File: server/controllers/user.js [hunk 4]
```diff
@@ -151,6 +150,83 @@ class User {
...
+      bio,
+      imageUrl,
+      createdAt,
+      updatedAt,
+    } = data;
+    return goodHttpResponse(
+      response,
+      200,
+      'User found',
+      request.isAuthorized ? data : {
```
Comment (seyi-adeleke) [inline server/controllers/user.js]: so if t request.isAuthorized is falsy? Also you shouldn't be doing this check in goodHttpResponse function

File: server/controllers/user.js [hunk 5]
```diff
@@ -151,6 +150,83 @@ class User {
...
+        bio,
+        imageUrl,
+        createdAt,
+        updatedAt,
+      },
+    );
+  }
+
+  /**
+   *
```
Comment (seyi-adeleke) [inline server/controllers/user.js]: this jsdoc isn't complete

File: server/controllers/user.js [hunk 6]
```diff
@@ -151,6 +150,83 @@ class User {
...
+        401,
+        'You are not permitted to complete this action',
+      );
+    }
+    if (request.body.images) {
+      const imageUrl = request.body.images[0];
+      request.body.imageUrl = imageUrl;
+    }
+    const updated = await userRepo.updateUser(request.body, request.params.username);
+    if (updated[0] === 1) {
```
Comment (seyi-adeleke) [inline server/controllers/user.js]: what's happening here? 🤔

File: server/middlewares/uploadImage.js [hunk 7]
```diff
@@ -0,0 +1,28 @@
+/* eslint prefer-const:0 */
+import cloudUpload from '../utilities/cloudUpload';
```
Comment (seyi-adeleke) [inline server/middlewares/uploadImage.js:1]: space after line 2

File: server/middlewares/uploadImage.js [hunk 8]
```diff
@@ -0,0 +1,28 @@
+/* eslint prefer-const:0 */
```
Comment (seyi-adeleke) [inline server/middlewares/uploadImage.js]: yo whats the point of using a linter if we're going to keep doing this 🤕

File: server/middlewares/uploadImage.js [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+  let store = [];
+
+  if (image) {
+    await cloudUpload(store, image);
+    request.body.images = store;
+    return next();
+  }
+  // For an array of images.
+  if (images) {
+    await images.map(oneImage => cloudUpload(store, oneImage));
```
Comment (seyi-adeleke) [inline server/middlewares/uploadImage.js]: `map` returns an array. If you're not using the return value it's cleaner to use a forEach

File: server/middlewares/validations.js [hunk 10]
```diff
@@ -48,6 +49,44 @@ class inputValidator {
...
+   * @param {object} next callback function
+   * @returns {object} http response object
+   */
+  static editProfile(request, response, next) {
+    const { email, username, password } = request.body;
+    if (username || email || password) {
+      return badHttpResponse(
+        response,
+        400,
+        'Email or username cannot be changed at this time.',
```
Comment (seyi-adeleke) [inline server/middlewares/validations.js]: why not?

File: server/repository/userRepository.js [hunk 11]
```diff
@@ -90,39 +128,50 @@ class UserRepository {
...
+    return oneUser;
+  }
+
+  /**
+   * Updates the user profile
+   * @param { object } newData
+   * @param { string } username
+   * @returns { boolean } true for successful update.
+   */
+  static async updateUser(newData, username) {
```
Comment (seyi-adeleke) [inline server/repository/userRepository.js:198]: So are we creating a different endpoint to update an admin user?

File: server/routes/users.js [hunk 12]
```diff
@@ -2,12 +2,17 @@ import { Router } from 'express';
 import User from '../controllers/user';
 import inputValidator from '../middlewares/validations';
 import checkAuth from '../middlewares/checkAuth';
```
Comment (seyi-adeleke) [inline server/routes/users.js]: checkAuth should be `isAuthenticated`. Please change it

File: server/utilities/cloudUpload.js [hunk 13]
```diff
@@ -0,0 +1,22 @@
+/* eslint no-unused-vars:0 */
```
Comment (seyi-adeleke) [inline server/utilities/cloudUpload.js]: 🤕

File: server/utilities/cloudUpload.js [hunk 14]
```diff
@@ -0,0 +1,22 @@
...
+import cloudinary from 'cloudinary';
+import cloudinaryConfig from '../config/cloudinaryConfig';
```
Comment (seyi-adeleke) [inline server/utilities/cloudUpload.js]: isn't this import unused?

File: server/utilities/userInput.js [hunk 15]
```diff
@@ -34,6 +34,18 @@ const validatePassword = (password) => {
   return passwordPattern.test(password);
 };
 
+/**
```
Comment (seyi-adeleke) [inline server/utilities/userInput.js:37]: This function doesn't actually confirm that the file is an image. I thought we discussed this?

**Review by seyi-adeleke — CHANGES_REQUESTED** [review]

File: server/middlewares/validations.js (hunk 10 above)
Comment (TheoOkafor) [reply server/middlewares/validations.js]: It was the team's recommendation when I asked them.

File: server/middlewares/uploadImage.js (hunk 9 above)
Comment (TheoOkafor) [reply server/middlewares/uploadImage.js]: Thanks

File: server/controllers/user.js (hunk 4 above)
Comment (TheoOkafor) [reply server/controllers/user.js]: It returns `{id,
        firstName,
        lastName,
        facebook,
        google,
        twitter,
        bio,
        imageUrl,
        createdAt,
        updatedAt,
      },`
The difference is that `data` has **email** and **isConfirmed**

Comment (TheoOkafor) [reply server/controllers/user.js]: Okay

File: server/repository/userRepository.js (hunk 11 above)
Comment (TheoOkafor) [reply server/repository/userRepository.js:198]: Not really.  We can use this endpoint.
Thanks.

File: package.json (hunk 3 above)
Comment (TheoOkafor) [reply package.json]: A mistake.

File: server/controllers/user.js (hunk 6 above)
Comment (TheoOkafor) [reply server/controllers/user.js]: Sequelize returns the number of rows that were changed. In this case, I was expecting just one row.

File: server/utilities/userInput.js (hunk 15 above)
Comment (TheoOkafor) [reply server/utilities/userInput.js:37]: Yes, we did.

Comment (TheoOkafor) [reply server/utilities/userInput.js:37]: I worked with Victor on that. He suggested that we can implement something on the frontend to do the confirmation.

File: server/utilities/cloudUpload.js (hunk 14 above)
Comment (TheoOkafor) [reply server/utilities/cloudUpload.js]: It is being used in the background by cloudinary.

File: server/utilities/userInput.js (hunk 15 above)
Comment (victor-nweze-deel) [reply server/utilities/userInput.js:37]: Yes, it turns out, after much pondering, that a backend mime check for images isn't possible since we are only storing the links in our database and not the actual file. So like Theo said, the checking will have to be on the front end

File: server/middlewares/uploadImage.js (hunk 7 above)
Comment (TheoOkafor) [reply server/middlewares/uploadImage.js:1]: Okay

File: server/utilities/userInput.js (hunk 15 above)
Comment (TheoOkafor) [reply server/utilities/userInput.js:37]: I believe cloudinary checks for that. So I have implemented a fail-safe for that using _try... catch_ method. I believe the edge case has been covered.

Comment (TheoOkafor) [reply server/utilities/userInput.js:37]: screen shot 2018-11-02 at 4 41 09 pm

File: server/repository/userRepository.js [hunk 16]
```diff
@@ -18,6 +18,7 @@ class UserRepository {
   }
 
   /**
+<<<<<<< HEAD
```
Comment (seyi-adeleke) [inline server/repository/userRepository.js]: please fix

File: server/repository/userRepository.js [hunk 17]
```diff
@@ -51,6 +52,9 @@ class UserRepository {
 
   /**
  * Finds a user by email
+=======
+ * Finds a user by email and username
+>>>>>>> feat(profile): Get and Update user profile
```
Comment (seyi-adeleke) [inline server/repository/userRepository.js]: here as well

File: server/repository/userRepository.js [hunk 18]
```diff
@@ -122,39 +164,49 @@ class UserRepository {
 
   /**
    * This method finds all non-admin users in the database
```
Comment (seyi-adeleke) [inline server/repository/userRepository.js]: Isnt this the same thing as line 84?

File: server/utilities/userInput.js (hunk 15 above)
Comment (seyi-adeleke) [reply server/utilities/userInput.js:37]: Ah, yeah you're right. This should be cloudinary's problem

**Review by seyi-adeleke — CHANGES_REQUESTED** [review]

File: server/repository/userRepository.js (hunk 16 above)
Comment (TheoOkafor) [reply server/repository/userRepository.js]: I missed that. Thanks.

File: server/repository/userRepository.js (hunk 18 above)
Comment (TheoOkafor) [reply server/repository/userRepository.js]: It is. I have changed that.

File: server/routes/users.js (hunk 12 above)
Comment (TheoOkafor) [reply server/routes/users.js]: okay. Thanks

**Review by seyi-adeleke — APPROVED** [review]
